### ISSUE TITLE
Add form sections to documentation

### ISSUE THREAD

**Issue by bmignano** [patternfly/patternfly-next#1831]
In OpenShift, we need the ability to box off fields in a form that make up a distinct elements within the form. We investigated using indentation as a solution, but the boxes create a more clear hierarchy and are more visually apparent. Attached are two (low-fidelity) mocks of where we would use this feature.

FYI @beanh66 @andybraren @cshinn 

![create-secret](https://user-images.githubusercontent.com/40238481/57883777-374d0c80-77f5-11e9-93e1-ce7459e6f648.png)
![create-network-policy-FORM](https://user-images.githubusercontent.com/40238481/57883781-3916d000-77f5-11e9-88ad-e1bbcde284ef.png)

**Comment by maryshak1996** [comment]
I'd be happy to do visuals for this :)

**Comment by mcarrano** [comment]
@mceledonia @kybaker would be interested to hear your thoughts on this.

**Comment by kybaker** [comment]
@mcarrano Seems like a valid usecase, but I do think we need to do some design exploration for this to think about the visuals and how it scales in the various layouts.

**Comment by mcarrano** [comment]
I agree @kybaker . I have tagged it as an enhancement request and we should consider for placement on the readmap. @LHinson @rachael-phillips FYI

**Comment by rachael-phillips** [comment]
@LHinson @mcarrano I've added this to the Unplanned section of our release planning spreadsheet for us to discuss in our next planning meeting.

**Comment by maryshak1996** [comment]
After some discussion with @kybaker @chart @bmignano and others on the OpenShift side, I think that I have a design that fits the use case. Simple visual spec to come soon :) 

**SketchCloud:** https://sketch.cloud/s/jP5QG?settings=jP5QG
**Marvel Prototype**: https://marvelapp.com/63g9616

**Comment by mcarrano** [comment]
This looks good to me @maryshak1996 .  @mcoker can you take a look at this? Would you see defining a new component that is simply a container for form content?

**Comment by mcoker** [comment]
@maryshak1996 @mcarrano Sorry for the delay, this looks good to me. It looks like an element of the form component. So just a part of the existing form component versus a new component, unless we want to abstract this out in a way where it can be used outside of the context of the form component.

**Comment by bmignano** [comment]
@maryshak1996 @mcarrano It was just pointed out to me that this design doesn't account for more complex form groups such as the second screenshot in the description above. In cases where you can add more elements to a form group (such as adding a Port in the screenshot), the `- Remove` link would conflict with the placement of the `+ Add another` link. Could we take another look at this design to account for these cases?

**Comment by mcarrano** [comment]
@bmignano From a PatternFly perspective, I'm not sure that we need anything new to implement those actions inside of a form section. It's just a matter of placing a link-style button within the contents of the form section. What that button does, including adding more elements to the form, should be within the applications control. @mcoker please verify to make sure I'm not saying something incorrect here :)  The most we might want to do is create a demo that embodies design best practices on link placement.

**Comment by mcoker** [comment]
@mcarrano agreed, the actions would just be put in another form row. Here's a quick mockup from our existing [form demo](https://pf4.patternfly.org/demos/BasicForms/examples/)

<img width="626" alt="Screen Shot 2019-07-30 at 2 34 54 PM" src="https://user-images.githubusercontent.com/35148959/62159426-56791780-b2d7-11e9-85c9-3f9bccbfd551.png">

<img width="633" alt="Screen Shot 2019-07-30 at 2 34 59 PM" src="https://user-images.githubusercontent.com/35148959/62159425-56791780-b2d7-11e9-8543-f12430b38088.png">

<img width="608" alt="Screen Shot 2019-07-30 at 2 35 06 PM" src="https://user-images.githubusercontent.com/35148959/62159424-56791780-b2d7-11e9-902e-f520ec163bb1.png">

**Comment by maryshak1996** [comment]
<img width="1039" alt="Screen Shot 2019-07-30 at 4 07 42 PM" src="https://user-images.githubusercontent.com/19824574/62161482-49aef080-b2e4-11e9-9870-e5f8e8a86648.png">
@mcarrano @mcoker I think that where the confusion is in cases where there is the idea of "nested form elements". for example, needing the ability to add another form element within a form section

**Comment by maryshak1996** [comment]
^ also note: I am wondering if some additional styling to the add rule might be helpful (styling that area as if it were another section box)

**Comment by mcoker** [comment]
@maryshak1996 ah, are you referring to the multiple form controls under "Ports"? If so, we don't support that, but should. I've mentioned adding support for that in the effort for https://github.com/patternfly/patternfly-next/issues/919. And in that case, I think the "add port" action would be another element in that form-group, and the "remove ingress rule" action would be another form-group in the section. Then "add another ingress" rule would just be a standalone action in its own form section.

**Comment by bmignano** [comment]
@mcoker Yep that's what we were referring to! In this case, do you think we need to reevaluate the placement of the `Remove ingress rule` to make the hierarchy of actions more clear? I think if we left the design as is, the screenshot @maryshak1996 posted above is accurate. From a design perspective, I was just wondering if those 3 add / remove links are confusing where they are.

**Comment by mcoker** [comment]
@bmignano I think it's OK the way it is, particularly with the buttons' wording in relation to the whole section's title "Ingress Rules." I wonder if "add another ingress rule" needs to be in a section (with the border/padding treatment) or should just be an action listed under the top section left-aligned with the left border of the section above it. Also if there should be more separation for "Remove ingress rule" - like a top border, or light grey background treatment? Not sure if those would be improvements or not, just things that come to mind looking at the design.

**Comment by bmignano** [comment]
@mcoker Thanks, makes sense to me! I'll let @maryshak1996 address the visual treatment :)

**Comment by maryshak1996** [comment]
@mcoker @bmignano 
I'll take another look at the remove section in particular, to separate that out from the other links more. I'll post some updated mockups tomorrow or friday!

**Comment by maryshak1996** [comment]
@bmignano @mcoker @mcarrano 
what do you guys think of something like this? it's a bit clunky so I could play around with spacing a bit more, but I'm curious if you guys feel like this satisfies the use case a little better?
![Screen Shot 2019-08-01 at 4 14 43 PM](https://user-images.githubusercontent.com/19824574/62324466-97a82d80-b477-11e9-84d9-8c3a62a035a2.png)

**Comment by mcarrano** [comment]
@maryshak1996 So if I read this right, it's just adding a horizontal divider insider of the frame, correct?  Seems OK to me. @mcoker to we have the ability to create horizontal dividers within forms now?

**Comment by mcoker** [comment]
We do, you can add a divider component between rows, but it wouldn't work out of the box the way I envision adding sections. By default I think the sections need padding to create space between the border and the content, so the divider wouldn't stretch to the edges of the section.

<img width="821" alt="Screen Shot 2019-08-01 at 4 02 25 PM" src="https://user-images.githubusercontent.com/35148959/62327330-f2408a00-b475-11e9-991a-303360c8188a.png">

If this is the ultimate design we want to go with, we can 1) write it in a way that a divider placed in a section would always extend to the edges of the section, 2) offer the divider in the above screenshot as the default and add a modifier to have the divider extend to the edge of the section, or 3) leave it up to the app developer to make that modification/override on their end if they want the divider to extend to the edge of the section.

**Comment by maryshak1996** [comment]
@mcoker I think that option 2 makes the most sense. with this modifier option, the designer could choose to use the standard section divider that doesn't span full width somewhere else in the section, but then use it with the modifier just for that bottom section, right?

**Comment by mcoker** [comment]
@maryshak1996 yep!

**Comment by maryshak1996** [comment]
@mcoker fantastic! I think that that sounds perfect then :)

**Comment by maryshak1996** [comment]
I'll update my mocks and spec right now, so the above links will have this new information in a minute

**Comment by maryshak1996** [comment]
Should be all updated now!

**Comment by mcarrano** [comment]
@mcoker @maryshak1996 I'm having a hard time identifying whether there is a follow-up action here. Can you take a look at this again? Unless there is something specific we should be adding to PatternFly, I am going to close.

**Comment by mcoker** [comment]
@mcarrano It looks like the task was to create sections/groups, which we don't currently have. There look to be other updates to the form to fully support this use case - stacking form control elements associated with a single label, for example, but that doesn't seem to be necessary for the section enhancement.

**Comment by mcarrano** [comment]
@maryshak1996 @bmignano Is this something that is still needed? If not I am going to close as this request is quite stale.

**Comment by bmignano** [comment]
@mcarrano Yes, we're still interested in this design! It sounds like there are a number of use cases across OpenShift, and I think a couple of other products too. I think some folks will be sharing some more examples here as well.

**Comment by lclay2** [comment]
@bmignano @mcarrano Sounds like something we might be able to use in Kogito as well, to group (and possibly nest) form sections. Relates to this jira: https://issues.redhat.com/browse/KOGITO-2223. cc @srambach

**Comment by mcarrano** [comment]
OK. Will transfer this to design and add to the backlog.
